include/ofi_net.h

```diff
@@ -522,6 +536,8 @@ struct ofi_bsock {
 	struct ofi_sockctx tx_sockctx;
 	struct ofi_sockctx rx_sockctx;
 	struct ofi_sockctx cancel_sockctx;
+	struct ofi_sockctx pollin_sockctx;
+	struct ofi_sockctx pollout_sockctx;
```

Comment:
We're up to 5 sockctx structs per socket now.  Does it make sense to allocate a pool of sockctx with the domain and grow that as needed?  Quickly searching the code, it looks like that may be a significant change, given how frequently uring_sqe_inuse is accessed.  I'm guessing we'd still need to keep that level of state, but maybe a single set of flags could track that instead?

Reply:
I need to think a bit more about that... The `ofi_sockctx` structure is composed of a boolean + a pointer to the `ofi_bsock` structure. When a CQE is polled, the completion function needs to resolve what `ofi_bsock` structure has triggered the completion, and the type of the completion event. The problem is that io_uring allows for a 8-byte context only :)

Reply:
Can the ofi_sockctx carry this information?  It makes sense for ofi_bsock to have the tx/rx contexts.  But if the others are used sparingly, a pool might make sense.

Reply:
> Can the ofi_sockctx carry this information?

What information you mean? `ofi_sockctx` contains all the information to figure out what operation (RX, TX, cancel, etc...) completed and for which FID. 

>  It makes sense for ofi_bsock to have the tx/rx contexts. But if the others are used sparingly, a pool might make sense.

Mixing tx/rx sockctxs from the ofi_bsock and sockctx from a dynamic pool for the others would significantly complicate the sockapi interface. I also tried to move all our sockctx into a bufpool but that also requires many changes to the sockapi interface.

Considering that canceling a request is a blocking operation, `cancel_sockctx` does not need to stored in the sockapi. It can just be allocated on the caller's stack. As for `pollin_sockctx`, I'm not sure it's effectively required: if we can keep a receive buffer always posted, we should never need to wait for a POLLIN event. It should be the case already, but I haven't be able to confirm it's always true.

Reply:
I don't want to complicate the implementation for a minor optimization, like saving a few bytes of data per ep. If 2 of the sockctx aren't needed and can be removed at some point, then we're okay.